source/extensions/common/tap/admin.cc

```diff
@@ -110,40 +109,201 @@ void AdminHandler::unregisterConfig(ExtensionConfig& config) {
   }
 }
 
-void AdminHandler::AdminPerTapSinkHandle::submitTrace(
-    TraceWrapperPtr&& trace, envoy::config::tap::v3::OutputSink::Format format) {
+PerTapSinkHandlePtr
+AdminHandler::createPerTapSinkHandle(uint64_t,
+                                     envoy::config::tap::v3::OutputSink::OutputSinkTypeCase type) {
+  using ProtoOutputSinkType = envoy::config::tap::v3::OutputSink::OutputSinkTypeCase;
+
+  /**
+   * Switching on the sink type here again after doing so in TapConfigBaseImpl constructor
+   * seems a bit strange. A possible refactor in the future could involve moving all Sinks
+   * to live where the FilePerTapSink lives, and passing in the sink to use into the AdminHandler.
+   */
+
+  // Select the sink implementation to use based on type specified in YAML request body
+  switch (type) {
+  case ProtoOutputSinkType::kStreamingAdmin:
+    return std::make_unique<AdminPerTapSinkHandle>(*this);
+  case ProtoOutputSinkType::kBufferedAdmin:
+    return std::make_unique<BufferedPerTapSinkHandle>(*this);
+  case ProtoOutputSinkType::kFilePerTap:
+  case ProtoOutputSinkType::kStreamingGrpc:
+    PANIC("not implemented");
+  case ProtoOutputSinkType::OUTPUT_SINK_TYPE_NOT_SET:
+  default:
+    PANIC_DUE_TO_CORRUPT_ENUM;
+  }
+}
+
+void AdminHandler::TraceBuffer::bufferTrace(
+    std::shared_ptr<envoy::data::tap::v3::TraceWrapper> trace) {
+  // Ignore traces once the buffer is full or flushed
+  if (flushed() || full()) {
+    return;
+  }
+
+  buffer_->emplace_back(std::move(*trace));
+}
+
+void AdminHandler::AdminPerTapSinkHandle::submitTrace(TraceWrapperPtr&& trace,
+                                                      envoy::config::tap::v3::OutputSink::Format) {
   ENVOY_LOG(debug, "admin submitting buffered trace to main thread");
   // Convert to a shared_ptr, so we can send it to the main thread.
   std::shared_ptr<envoy::data::tap::v3::TraceWrapper> shared_trace{std::move(trace)};
+  // Non owning pointer to the attached request, does not preserve lifetime unless in use
+  std::weak_ptr<AttachedRequest> weak_attached_request(parent_.attached_request_);
+
   // The handle can be destroyed before the cross thread post is complete. Thus, we capture a
   // reference to our parent.
-  parent_.main_thread_dispatcher_.post([&parent = parent_, trace = shared_trace, format]() {
-    if (!parent.attached_request_.has_value()) {
+  parent_.main_thread_dispatcher_.post([weak_attached_request, shared_trace] {
+    // Take temporary ownership - extend lifetime
+    std::shared_ptr<AttachedRequest> attached_request = weak_attached_request.lock();
+    if (!attached_request) {
+      ENVOY_LOG(debug, "attached request does not exist, not streaming trace");
```

Comment:
should we do something here with the failed request, like emit an HTTP status, maybe 408 or 410, and maybe detail in the response buffer also?

Reply:
I would leave this is as is for a couple reasons. First, I don't think we will have anywhere to send the HTTP response code:
- The response code is already sent at this point because `submitTrace` is called by envoy worker threads, and an HTTP response code is sent when `AdminHandler::handler` completes. 
- Without the attached request, we don't know where to send any of this information

Second, Depending on when the destruction of `attached_request_` occurs, this case may be hit somewhat frequently. For example take the following sequence of events:
1. Worker thread 1 calls `submitTrace` and makes a post to the main thread
2. Worker thread 2 calls `submitTrace` and makes a post to the main thread
3. Main thread executes the post from Worker 1 and finds the buffer is full. The buffer flushes and the stream is closed server side.
4. The onDestroyCallback for the admin stream runs
5. The attached request is deleted
6. The pending post from worker 2 runs, now `admin_request_` does not exist but nothing really went wrong

Reply:
For sure, I'll add some comments with this info